crypto/pkcs7/pkcs7.c

```diff
@@ -1409,3 +1415,246 @@ int PKCS7_decrypt(PKCS7 *p7, EVP_PKEY *pkey, X509 *cert, BIO *data, int flags) {
+    goto err;
+  }
+  return 1;
+err:
+  OPENSSL_PUT_ERROR(X509, ERR_R_CRYPTO_LIB);
+  return 0;
+}
+
+static int pkcs7_x509_add_certs_new(STACK_OF(X509) **p_sk,
+                                    STACK_OF(X509) *certs) {
+  GUARD_PTR(p_sk);
+  if (!certs) {  // |certs| can be null in the caller
+    return 1;
+  }
+  for (size_t i = 0; i < sk_X509_num(certs); i++) {
+    if (!pkcs7_x509_add_cert_new(p_sk, sk_X509_value(certs, i)))
+      return 0;
+  }
+  return 1;
+}
+
+static int pkcs7_signature_verify(BIO *in_bio, PKCS7 *p7, PKCS7_SIGNER_INFO *si,
+                                  X509 *signer) {
+  GUARD_PTR(in_bio);
+  GUARD_PTR(p7);
+  GUARD_PTR(si);
+  GUARD_PTR(signer);
+  EVP_MD_CTX *mdc_copy = NULL;
+  int ret = 0;
+
+  const int md_type = OBJ_obj2nid(si->digest_alg->algorithm);
+  EVP_MD_CTX *mdc = NULL;
+  BIO *bio = in_bio;
+  // There may be multiple MD-type BIOs in the chain, so iterate until we find
+  // the BIO with MD type we're looking for.
+  while (bio) {
+    if ((bio = BIO_find_type(bio, BIO_TYPE_MD)) == NULL) {
+      OPENSSL_PUT_ERROR(PKCS7, PKCS7_R_UNABLE_TO_FIND_MESSAGE_DIGEST);
+      goto err;
+    }
+    if (!BIO_get_md_ctx(bio, &mdc) || !mdc) {
+      OPENSSL_PUT_ERROR(PKCS7, ERR_R_PKCS7_LIB);
+      goto err;
+    }
+    if (EVP_MD_CTX_type(mdc) == md_type) {  // found it!
+      break;
+    }
+    bio = BIO_next(bio);
+  }
+
+  // Make a copy of |mdc| so we don't finalize the MD_CTX owned by |bio|
+  if ((mdc_copy = EVP_MD_CTX_new()) == NULL ||
+      !EVP_MD_CTX_copy_ex(mdc_copy, mdc)) {
+    OPENSSL_PUT_ERROR(PKCS7, ERR_R_EVP_LIB);
+    goto err;
+  }
+
+  // We don't currently support signed attributes. See |PKCS7_NOATTR|.
+  if (si->auth_attr && sk_X509_ATTRIBUTE_num(si->auth_attr) != 0) {
+    OPENSSL_PUT_ERROR(PKCS7, PKCS7_R_INVALID_SIGNED_DATA_TYPE);
+    goto err;
+  }
+
+  EVP_PKEY *pkey;
+  if ((pkey = X509_get0_pubkey(signer)) == NULL) {
+    goto err;
+  }
+
+  ASN1_OCTET_STRING *data_body = si->enc_digest;
+  if (!EVP_VerifyFinal(mdc_copy, data_body->data, data_body->length, pkey)) {
+    OPENSSL_PUT_ERROR(PKCS7, PKCS7_R_SIGNATURE_FAILURE);
+    goto err;
+  }
+
+  ret = 1;
+
+err:
+  EVP_MD_CTX_free(mdc_copy);
+  return ret;
+}
+
+int PKCS7_verify(PKCS7 *p7, STACK_OF(X509) *certs, X509_STORE *store,
+                 BIO *indata, BIO *out, int flags) {
+  GUARD_PTR(p7);
+  GUARD_PTR(store);
+  STACK_OF(X509) *signers = NULL, *untrusted = NULL;
+  X509_STORE_CTX *cert_ctx = NULL;
+  BIO *p7bio = NULL;
+  int ret = 0;
+
+  if (!PKCS7_type_is_signed(p7)) {
+    OPENSSL_PUT_ERROR(PKCS7, PKCS7_R_WRONG_CONTENT_TYPE);
+    goto out;
+  }
+
+  // If |p7| is detached, caller must supply data via |indata|
+  if (PKCS7_is_detached(p7) && indata == NULL) {
+    OPENSSL_PUT_ERROR(PKCS7, PKCS7_R_NO_CONTENT);
+    goto out;
+  }
+
+  // We enforce OpenSSL's PKCS7_NO_DUAL_CONTENT flag in all cases for signed
+  if (PKCS7_type_is_signed(p7) && (!PKCS7_is_detached(p7) && indata)) {
```

Comment:
It was already determined that `(PKCS7_type_is_signed(p7)` is non-zero on line 1569 above.  This condition could simply be `if (!PKCS7_is_detached(p7) && indata) {...}`.

Reply:
This is vestigial from when I tried to shoe-horn support for signedAndEnveloped into `PKCS7_verify`. yes, we can remove it now.